content/blog/2023-10-18-skywalking-toolkit-trace/index.md

```diff
@@ -0,0 +1,285 @@
+}
+
+func main() {
+	http.HandleFunc("/provider", providerHandler)
+
+	_ = http.ListenAndServe(":9998", nil)
+}
+
+```
+
+
+
+Then, in the terminal, execute:
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o consumer ./consumer.go
+./consumer
+```
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o provider ./provider.go
+./provider
+```
+
+```shell
+curl 127.0.0.1:9999/consumer
+```
+
+At this point, the UI will display the span information you created.
+
+![Trace List](./trace.png)
+
+If you need to trace methods that are executed only locally, you can use trace.CreateLocalSpan(). If you don't need to monitor information or states from the other end, you can change ExitSpan and EntrySpan to LocalSpan.
+
+The usage examples provided are for illustration purposes, and users can decide the tracing granularity and where in the program they need tracing.
+
+Please note that if a program ends too quickly, it may cause tracing data to be unable to be asynchronously sent to the SkyWalking backend.
+
+## Populate The Span
+
+When there's a necessity to record additional information, including creating/updating tags, appending logs, and setting a new operation name of the current traced Span, these APIs should be considered. These actions are used to enhance trace information, providing a more detailed and precise contextual description, which aids in better understanding the events or operations being traced.
+
+Toolkit trace APIs provide a convenient way to access and manipulate trace data, including:
+
+- Setting Tags: `SetTag()`
+- Adding Logs: `AddLog()`
+- Setting Span Names: `SetOperationName()`
+- Getting various IDs: `GetTraceID()`, `GetSegmentID()`, `GetSpanID()`
+
+For example, if you need to record the HTTP status code in a span, you can use the following interfaces while the span is not yet finished:
+
+```go
+trace.CreateExitSpan("GET:/provider", "localhost:9999", func(headerKey, headerValue string) error {
+	r.Header.Add(headerKey, headerValue)
+	return nil
+})
+resp, err := http.Get("http://localhost:9999/provider")
+trace.SetTag("status_code", fmt.Sprintf("%d", resp.StatusCode))
+spanID := trace.GetSpanID()
+trace.StopSpan()
+```
+
+It's important to note that when making these method calls, the current thread should have an active span.
+
+## Async APIs
+
+Async APIs work for manipulating spans across Goroutines. These scenarios might include:
+
+- Applications involving concurrency or multiple goroutines where operating on Spans across different execution contexts is necessary.
+- Updating or logging information for a Span during asynchronous operations.
+- Requiring a delayed completion of a Span.
+
+To use it, follow these steps:
+
+- Obtain the return value of CreateSpan, which is `SpanRef`.
+- Call `spanRef.PrepareAsync()` to prepare for operations in another goroutine.
+- When the current goroutine's work is done, call `trace.StopSpan()` to end the span (affecting only in the current goroutine).
```

Comment:
What happens if I pass the span ref to multiple go routines, does each one of them need to call `AsyncFinish()`?

Reply:
No, you just need to call the `AsyncFinish` in one goroutine. If you call it multiple times, it would throw an error, right @Alipebt?

Reply:
Yes, it will throw an error about having closed it